### PR TITLE
ARTEMIS-3604 Async sends overflowing server in OpenWire

### PR DESCRIPTION
**Pull request by clebertsuconic**

### CODE REVIEW

File: artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireConnection.java
Comment (franz1981) [inline artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireConnection.java:242]: I would use an actor which queue is bounded (we have many in JCtools) adding a method boolean tryAct
that return false if the queue has reached maximum capacity.
This would be more performant and easy to manage, given that could be used by other part of code to limit the size of tasks on the actor.

**Review by franz1981 — CHANGES_REQUESTED** [review]

File: artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireProtocolManager.java
Comment (gtully) [inline artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireProtocolManager.java:103]: the actor is in units of commands, I wonder if it would be possible to limit by bytes size and match to the configured serverSocketBuffer size. I guess that raw bytes are already unmarshalled when we are dealing with commands.
It is just that it seems natural to map the pending data on in the broker to what can be pending on the tcp layer. So from a client perspective there is max 2*socketBufferSize bytes pending. Or maybe it is 3x, with one being on the client side of the tcp buffer.

Comment (clebertsuconic) [reply artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireProtocolManager.java:103]: I wouldn't know how to do that TBH

Comment (clebertsuconic) [reply artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireProtocolManager.java:103]: I was going with 1000, but my computer is slow to sync due to some crypto thing...


I have not decided on the default yet.. but this is mostly about syncs...


the other thing is I will probably make those writes to the journal to not sync if the client is not waiting for a response.

File: artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireConnection.java
Comment (clebertsuconic) [reply artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireConnection.java:242]: the thing is, I may have to go beyond the limit some times.. because the parse on the packet already happened. and the messages were already read from the network. I have to read them all.

File: artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireConnection.java
Comment (gtully) [inline artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireConnection.java:357]: should this no go through checkMemory - to sync up with the page store blocking. If the page store is still blocked we don't want to enable auto read till there is some space.

I am thinking, we don't have to be careful about disabling auto read, b/c we will aways get to do it again... but with enabling, having an empty work queue is not sufficient, there also has to be space. so we need to coordinate the enabling.

Comment (clebertsuconic) [reply artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireConnection.java:341]: @gtully when blocked, autoRead will be false.

We are still calling the flush o this case though.. 

we will need the paging to send an event to set this call.

File: artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireProtocolManager.java
Comment (gtully) [inline artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireProtocolManager.java:243]: this does not make sense to me, a config that should not be used. The intent is to bound the queue and it is matched to the receive buffer size. if it needs to be lower or higher then modify the acceptor receive buffer size. The bounding of the actor stops reading, so at a max the receive buffer size will be pending on the socket and in the broker. Then the client will block on a tcp write. When will it make sense to have more than send buffer pending on the broker? I don't know how to reason about setting a maxActorSize on the protocol manager!
in terms of a name, it could be actorPendingQueueMaxSizeBytes but that is an aside. Only add that if we know when it should be set.

File: artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireConnection.java
Comment (gtully) [inline artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireConnection.java:239]: this seems the only right option to me, if the actor can be used without an acceptor, then maybe pass in the maxPendingSize rather than the acceptorUsed.

File: artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireConnection.java
Comment (gtully) [inline artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireConnection.java:820]: did a quick check, the only way getSize would return a different value after act is if org.apache.activemq.command.ActiveMQMessage#clearProperties was called and from what I can see we don't do that so this is a reasonable estimate. 
The 1k for other commands may be a little heavy, but I guess it is better to error on the side of caution. 
nice!

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java
Comment (franz1981) [inline artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java:36]: `private final` on the members

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java
Comment (franz1981) [inline artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java:54]: `scheduledFlush` must be `0` after `clearLimit::run` or a racy `act` after this would call `overLimit::run` before it

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java
Comment (franz1981) [inline artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java:66]: `sizeGetter.applyAsInt((T)message)` should check if the estimated size is > 0

**Review by franz1981 — CHANGES_REQUESTED** [review]
Generally looking good, but I'm still not sold on the bound actor abstraction because:

- `act` can be called by several threads concurrently while OpenWire does is single-thread, hence the way it behave isn't the same
- I cannot still fit in my mind what `flush` means and why it call `overLimit::run` regardless current actor pending size

I'm looking how the reactive stream abstraction is implemented in other systems because we probably need a backpressure advertizing with an overflow strategy that can accumulate beyond expected capacity and maybe there's already something implemented/tested around

File: artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireConnection.java
Comment (clebertsuconic) [reply artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireConnection.java:239]: this will be null on tests only...

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java
Comment (clebertsuconic) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java:66]: why?

File: artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireProtocolManager.java
Comment (gemmellr) [reply artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireProtocolManager.java:248]: Also, maxActorSize feels off as a name given it specifically isnt actually a max and can be exceeded, and indeed doesnt cause anything to happen until it has already been exceeded.

Again, not sure on great alternatives though...actorSizeThreshold?

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java
Comment (franz1981) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java:66]: because if not (a bug/issue on the implementation), it risks to not work as expected

File: artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireProtocolManager.java
Comment (clebertsuconic) [reply artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireProtocolManager.java:243]: @gtully if you have a slow device to sync, and you read say 1M worth of data, but take more than the configured TTL to read again your clients will start to fall apart.

This is a backup plan in case this scenario happens. By default it should use the tcp-read-buffer, but when things fall apart you have the option to use a lower read-value. I have seen that happening on my tests and I don't want to take the risk without having a backup plan.

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java
Comment (clebertsuconic) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java:66]: You're right.. this was a copy & paste from another place where I had it as an object ... lol

thanks

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java
Comment (clebertsuconic) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java:24]: @gemmellr SizeAwareActor?... I don't have a better name.. if you find one I will use it

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java
Comment (clebertsuconic) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java:66]: I think it would be ok to have a packet with zero weight.. if the user has a control or something. 
I don't consider this a bug.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java
Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java:717]: I don't have time to handle this one now.. I have to finish this PR today.

File: tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/paging/FloodServerWithAsyncSendTest.java
Comment (clebertsuconic) [reply tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/paging/FloodServerWithAsyncSendTest.java:78]: > Generally looking good, but I'm still not sold on the bound actor abstraction because:
> 
> * `act` can be called by several threads concurrently while OpenWire does is single-thread, hence the behaviour in a general purpose context won't be the same
> * I cannot still fit in my mind what `flush` means and why it call `overLimit::run` regardless current actor pending size
> 
> I'm looking how the reactive stream abstraction is implemented in other systems because we probably need a backpressure advertizing mechanism with an overflow strategy that can accumulate beyond expected capacity and maybe there's already something implemented/tested around

this is about stop reading after a certain limit is over. I still need to parse everything that's on the buffer.. I can't just limit the actor here... things would break otherwise.

Comment (franz1981) [reply tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/paging/FloodServerWithAsyncSendTest.java:78]: Yeah, in this case is a limit aware actor in this case: it better capture what it does

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java
Comment (franz1981) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java:66]: Yes but negative is a problem

**Review by franz1981 — CHANGES_REQUESTED** [review]

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java
Comment (franz1981) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java:54]: @clebertsuconic this must be change to save bad things to happen concurrently

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java
Comment (gemmellr) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java:24]: Still not great, but definitely better than 'Bound' given its not actually bound.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java
Comment (gemmellr) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java:717]: Its unclear to me that you can trust this PR without changing it, given the fact it would be running tasks when it shouldnt, and these are now coordinating on important state.

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java
Comment (clebertsuconic) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java:24]: I'm naming it ThresholdAwareActor

Comment (clebertsuconic) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java:24]: ThresholdActor actually, I'm not using the Aware word

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java
Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java:717]: I will address this on a next PR if there's an issue here.

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java
Comment (gemmellr) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java]: Its better than Bound :)

File: tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/paging/FloodServerWithAsyncSendTest.java
Comment (gemmellr) [reply tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/paging/FloodServerWithAsyncSendTest.java:80]: Your right, it isnt run in the fast-tests profile..I thought it was, but its not...its just burning time building the module and generating things, though it doesnt actually run them. I'd still probably put it elsewhere.

File: tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/paging/FloodServerWithAsyncSendTest.java
Comment (clebertsuconic) [reply tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/paging/FloodServerWithAsyncSendTest.java:121]: The point is that a third connection should still allowed to perform. The server should not become unresponsive which is what's happening on this situation.

My laptop has some bad cryptographic thing that slows the server down to a point where I can't guarantee the receives and senders will actually happen. I thought it would be safer to just check the unresponsive server than the actual producers and consumers.

the test is reproducing the issue as I expected here.

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java
Comment (clebertsuconic) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java]: it's a single thread setting this to false.. there's no concurrency here.

Comment (franz1981) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java]: `act` can set compareAndSet(0, 1) right after this `this.scheduledFlush = 0` meaning that the Runnable will run without the right order

Comment (clebertsuconic) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java]: Right... the flush should set it to 0 no matter what.

Comment (franz1981) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java]:
```
tasks = {1, 2, 3, F}
size = 3
max_size = 2
schedule_flush = 1

act(4)				|	actor's doTask) 	
--------------------------------|-------------------------------
size: 3 -> 4			|				
				|	tasks = {1, 2, 3, F} - {F}
				|	schedule_flush: 0
				|	
				|				
schedule_flush: 0 -> 1		|				
				|				
overLimit.run()			|				
				|	clearThreshold.run();
tasks = {} + {F}		|	
tasks = {} + {4}		|	
				|	tasks = {1, 2, 3} - {3}
				|	size: 4 -> 3
				```
				
				this is to help the future me on this :D

Comment (clebertsuconic) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java]: that's actually why I set it to 0 before calling...

I may eventually have two in flight if I get one packet arriving with more size then the whole actor...


If you think there's a bug in there feel free to send a new PR *after* this one is merged.. I need to merge this asap.

Comment (franz1981) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java]: No bud this is a bug, because that listeners calls are not supposed to run in random order, but in the right sequence

Comment (clebertsuconic) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java]: @franz1981 Ok, I am merging this now, If you want to improve this.. or talk to me next week.

Comment (franz1981) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java]: Ok, but I don't understand why you don't consider it a bug: listeners must not be called concurrently right? The compare and set is there to prevent it

Comment (clebertsuconic) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java]: I added a test for this condition.. I don't think I would have caused any harm on the openwire flow control... I'm just trying to fix the flow control in openwire now.

Comment (franz1981) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java]: I know but given that is a racy behaviour it won't happen easily, like any concurrent bugs, but the execution I shown is valid and can still happen so worth to fix it.
At worst it would still fix your issue: run your existing tests and if it pass it means that is still working as you expect, if not, feel free to merge it as is now and I'll find out later what can be wrong (if any).

Comment (clebertsuconic) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java]: I believe its fixed.

**Pull request merged**
